File: dailyproblem/09_03_2025/09_03_2025.cpp
// class Solution {
//     public:
//         int numberOfAlternatingGroups(vector<int>& colors, int k) {
//             int n = colors.size();
//             int count = 0;
    
//             // Check all valid windows
//             for (int i = 0; i < n; i++) {
//                 bool isAlternating = true;
//                 for (int j = 0; j < k - 1; j++) {
//                     if (colors[(i + j) % n] == colors[(i + j + 1) % n]) {
//                         isAlternating = false;
//                         break;
//                     }
//                 }
//                 if (isAlternating) count++;
//             }
    
//             return count;
//         }
//     };
    

class Solution {
    public:
        int numberOfAlternatingGroups(vector<int>& colors, int k) {
            int n = colors.size();
            vector<bool> is_alternating(n, false);
    
            // Precompute alternation check for each adjacent pair
            for (int i = 0; i < n; i++) {
                is_alternating[i] = (colors[i] != colors[(i + 1) % n]);
            }
    
            // Count valid k-length alternating groups using a sliding window
            int count = 0, valid_count = 0;
    
            // Initialize the first window
            for (int i = 0; i < k - 1; i++) {
                valid_count += is_alternating[i];
            }
    
            // Slide the window across the circular array
            for (int i = 0; i < n; i++) {
                if (valid_count == k - 1) count++;  // Valid alternating group found
                
                // Move the window forward
                valid_count -= is_alternating[i];
                valid_count += is_alternating[(i + k - 1) % n];
            }
    
            return count;
        }
    };
    

File: dailyproblem/09_03_2025/09_03_2025.py
# class Solution:
#     def numberOfAlternatingGroups(self, colors: List[int], k: int) -> int:
#         n = len(colors)
#         count = 0

#         for i in range(n):
#             is_alternating = True
#             for j in range(k - 1):
#                 if colors[(i + j) % n] == colors[(i + j + 1) % n]:
#                     is_alternating = False
#                     break
#             if is_alternating:
#                 count += 1
        
#         return count


from typing import List

class Solution:
    def numberOfAlternatingGroups(self, colors: List[int], k: int) -> int:
        n = len(colors)
        is_alternating = [colors[i] != colors[(i + 1) % n] for i in range(n)]

        count = 0
        valid_count = sum(is_alternating[:k - 1])  # Initial window

        # Sliding window across the circular array
        for i in range(n):
            if valid_count == k - 1:
                count += 1  # Valid alternating group found

            # Slide the window
            valid_count -= is_alternating[i]
            valid_count += is_alternating[(i + k - 1) % n]

        return count
